### PR TITLE
Changes in test tool and change raw pointer to pointer argument to cJSON

### LINKED ISSUE #5289: Begin and end virtual transaction
## Description

That a logical transaction is kept in memory, during the life cycle of a scan 
or other actions that are executed in multiple steps and by elements, in order to interact, alter the data to be stored, in other words the context is instantiated and deleted with these methods.

element sync (1)

## Testing

This implementation should be tested at the unit and integration level to ensure error and response handling.
integration tests must be done with dbsync_test_tool

## Tasks  / DoD
- [X] Implementation
- [X] Unit tests
- [X] CPPCHECK
- [X] Integration tests
- [X] ASAN
- [X] Valgrind
- [X] PR
- [X] PR Integration tests

### PR DESCRIPTION
**Pull request by Dwordcito**
|Related issue|
|---|
|5289|

##Issue description
That a logical transaction is kept in memory, during the life cycle of a scan
or other actions that are executed in multiple steps and by elements, in order to interact, alter the data to be stored, in other words, the context is instantiated and deleted with these methods.

##Description
- Changes in test tools to be able to execute generic actions.
- Change the approach from P2P to cJSON for tables argument in dbsync_create_txn

Closes #5289 

##Coverage
image

##Unit tests
image
image
image

image -> Flaky @konangc87 

##Valgrind
image

##CPPCHECK
image

### CODE REVIEW

File: src/dbsync/testtool/action.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+{
+    void operator()(char* json)
+    {
+        cJSON_free(json);
+    }
+};
+
+struct CJsonDeleter
+{
+    void operator()(cJSON* json)
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:26]: you can define both in the same struct, cJSON* and char* are different types.

File: src/dbsync/testtool/action.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+};
+
+struct IAction
+{
+    virtual void execute(std::unique_ptr<TestContext>& ctx, const nlohmann::json& value) = 0;
+};
+
+struct UpdateWithSnapshotAction : public IAction
+{
+    virtual void execute(std::unique_ptr<TestContext>& ctx, const nlohmann::json& value) override
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:48]: virtual not needed.

File: src/dbsync/testtool/action.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            const std::unique_ptr<cJSON, ResultDeleter> snapshotLambdaPtr(snapshotLambda);
+            
+            std::stringstream oFileName;
+            oFileName << "action_" << ctx->currentId << ".json";
+            const std::string outputFileName{ ctx->outputPath +"/"+oFileName.str() };
+
+            std::ofstream outputFile{ outputFileName };
+            const std::unique_ptr<char, CharDeleter> snapshotDiff{ cJSON_Print(snapshotLambdaPtr.get()) };
+            outputFile << snapshotDiff.get() << std::endl;
+            outputFile.close();
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:67]: close not needed.

File: src/dbsync/testtool/action.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    }
+};
+
+
+struct CreateTransactionAction : public IAction
+{
+    virtual void execute(std::unique_ptr<TestContext>& ctx,
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:75]: virtual not needed

File: src/dbsync/testtool/action.h [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            cJSON_Parse(value["body"]["tables"].dump().c_str())
+        };
+
+        ctx->txnContext = dbsync_create_txn(ctx->handle,
+                                     jsonTables.get(),
+                                     0,
+                                     0,
+                                     (result_callback_t)0x1);
+
+            std::stringstream oFileName;
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:89]: identation.

File: src/dbsync/testtool/action.h [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            oFileName << "action_" << ctx->currentId << ".json";
+            const auto outputFileName{ ctx->outputPath + "/" + oFileName.str() };
+
+            std::ofstream outputFile{ outputFileName };
+            const nlohmann::json jsonResult = { {"txn_context", nullptr != ctx->txnContext ? true : false } };
+            outputFile << jsonResult.dump() << std::endl;
+            outputFile.close();
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:96]: close not needed.

File: src/dbsync/testtool/action.h [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+{
+    virtual void execute(std::unique_ptr<TestContext>& ctx, 
+                         const nlohmann::json& /*value*/) override
+    {
+        
+
+        const auto retVal { dbsync_close_txn(ctx->handle,
+                                     ctx->txnContext)} ;
+
+            std::stringstream oFileName;
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:110]: identation,

File: src/dbsync/testtool/action.h [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            std::ofstream outputFile{ outputFileName };
+            const nlohmann::json jsonResult = { {"txn_context", nullptr != ctx->txnContext ? true : false } };
+            outputFile << jsonResult.dump() << std::endl;
+            outputFile.close();
+    }
+};
+
+struct CloseTransactionAction : public IAction
+{
+    virtual void execute(std::unique_ptr<TestContext>& ctx,
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:102]: virtual not needed.

File: src/dbsync/testtool/action.h [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                                     ctx->txnContext)} ;
+
+            std::stringstream oFileName;
+            oFileName << "action_" << ctx->currentId << ".json";
+            const auto outputFileName{ ctx->outputPath + "/" + oFileName.str() };
+
+            std::ofstream outputFile{ outputFileName };
+            const nlohmann::json jsonResult = { {"txn_context", retVal } };
+            outputFile << jsonResult.dump() << std::endl;
+            outputFile.close();
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:117]: close not needed.

File: src/dbsync/testtool/factoryAction.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Wazuh DBSYNC
+ * Copyright (C) 2015-2020, Wazuh Inc.
+ * June 21, 2020.
```
Comment (konangc87) [inline src/dbsync/testtool/factoryAction.h:4]: wrong date

File: src/dbsync/testtool/testContext.h [hunk 11]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Wazuh DBSYNC
+ * Copyright (C) 2015-2020, Wazuh Inc.
+ * June 21, 2020.
```
Comment (konangc87) [inline src/dbsync/testtool/testContext.h:4]: wrong date

File: src/dbsync/testtool/action.h [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+struct ResultDeleter
+{
+    void operator()(cJSON* json)
+    {
+        dbsync_free_result(&json);
+    }
+};
+
+struct IAction
```
Comment (konangc87) [inline src/dbsync/testtool/action.h:41]: virtual dtor missing.

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 13]
```diff
@@ -47,10 +47,10 @@ class CmdLineArgs
                   << "Options:\n"
                   << "\t-h \t\t\tShow this help message\n"
                   << "\t-c JSON_CONFIG_FILE\tSpecifies the json config file to initialize the database.\n"
-                  << "\t-s SNAPSHOT_LIST\tSpecifies the list of snapshots to exercise the dabase.\n"
+                  << "\t-a ACTION_LIST\tSpecifies the list of actions to exercise the database.\n"
```
Comment (tinachoB) [inline src/dbsync/testtool/cmdArgsHelper.h:50]: Readme.md file should be updated with this too.

File: src/dbsync/testtool/action.h [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+                                     0,
+                                     0,
+                                     (result_callback_t)0x1);
+
+            std::stringstream oFileName;
+            oFileName << "action_" << ctx->currentId << ".json";
+            const auto outputFileName{ ctx->outputPath + "/" + oFileName.str() };
+
+            std::ofstream outputFile{ outputFileName };
+            const nlohmann::json jsonResult = { {"txn_context", nullptr != ctx->txnContext ? true : false } };
```
Comment (tinachoB) [inline src/dbsync/testtool/action.h:94]: "? true : false" can be avoided here. Like this:
"const nlohmann::json jsonResult = { {"txn_context", nullptr != ctx->txnContext } };"

File: src/dbsync/testtool/main.cpp [hunk 15]
```diff
@@ -74,41 +48,30 @@ int main(int argc, const char* argv[])
...
             }
             dbsync_teardown();
-            std::cout << "Resulting files are located in the "<< cmdLineArgs.outputFolder() << "folder" << std::endl;
+            std::cout << "Resulting files are located in the " << cmdLineArgs.outputFolder() << " folder" << std::endl;
         }
         else
         {
-            std::cout << "\nSomething went wrong configuring the database. Please, check the config file data" << std::endl;
-            CmdLineArgs::showHelp();
+            std::cout << std::endl << "Something went wrong configuring the database. Please, check the config file data" << std::endl;
```
Comment (tinachoB) [inline src/dbsync/testtool/main.cpp:70]: I'm not seeing the "showHelp()" message called. Is this method still required?

**Pull request merged**